File: zero-server/src/main/kotlin/com/zeroday/handler/handlers/TaskHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.reqStr
import com.zeroday.model.TaskInstance
import com.zeroday.model.toMap
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import kotlinx.serialization.json.JsonObject

/**
 * Contract / task lifecycle: listing available contracts, accepting them
 * (turning an EnhancedContract into a TaskInstance on the player), and
 * marking a task as completed.
 */
class TaskHandler : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.GET_TASKS,
        RequestTypes.ACCEPT_TASK,
        RequestTypes.COMPLETE_TASK
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.GET_TASKS -> getTasks(ctx, playerId)
            RequestTypes.ACCEPT_TASK -> acceptTask(ctx, playerId, payload)
            RequestTypes.COMPLETE_TASK -> completeTask(ctx, playerId, payload)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported task request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun getTasks(ctx: HandlerContext, playerId: String): HandlerResult {
        val tasks = ctx.services.taskService.getAvailableTasks(playerId)
        return HandlerResult.Ok(ResponseTypes.TASKS, mapOf("available_tasks" to tasks))
    }

    private suspend fun acceptTask(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val taskId = try { payload.reqStr("task_instance_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Task instance ID required", ErrorCategory.VALIDATION))
        }
        return ctx.services.taskService.acceptTask(playerId, taskId).fold(
            onSuccess = { task -> HandlerResult.Ok(ResponseTypes.TASK_ACCEPTED, mapOf("task" to task)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Could not accept task", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun completeTask(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val taskId = try { payload.reqStr("task_instance_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Task instance ID required", ErrorCategory.VALIDATION))
        }
        return ctx.services.taskService.completeTask(playerId, taskId).fold(
            onSuccess = { task ->
                HandlerResult.Ok(
                    type = ResponseTypes.TASK_COMPLETED,
                    payload = mapOf(
                        "task" to task.toMap(),
                        "player" to ctx.services.playerService.getSnapshot(playerId)?.toMap()
                    )
                )
            },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Could not complete task", ErrorCategory.INTERNAL)) }
        )
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/TaskHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class TaskHandler : IHandler
{
    private readonly TaskService _taskService;
    private readonly IPlayerService _playerService;

    public string MessageType => "get_tasks";

    public TaskHandler(TaskService taskService, IPlayerService playerService)
    {
        _taskService = taskService;
        _playerService = playerService;
    }

    public async Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        if (payload is null)
            return new MessageResult("error", connectionId, new { message = "Missing payload" });

        var type = payload.Value.Str("type") ?? "get_tasks";

        return type switch
        {
            "get_tasks" => await GetTasks(connectionId),
            "accept_task" => await AcceptTask(connectionId, payload.Value),
            "complete_task" => await CompleteTask(connectionId, payload.Value),
            _ => new MessageResult("error", connectionId, new { message = $"Unsupported task request: {type}" })
        };
    }

    private async Task<IActionResult> GetTasks(string connectionId)
    {
        var tasks = await _taskService.GetAvailableTasks(connectionId);
        return new MessageResult("tasks", connectionId, new { available_tasks = tasks });
    }

    private async Task<IActionResult> AcceptTask(string connectionId, JsonElement payload)
    {
        string taskInstanceId;
        try
        {
            taskInstanceId = payload.ReqStr("task_instance_id");
        }
        catch
        {
            return new MessageResult("error", connectionId, new { message = "Task instance ID required" });
        }

        var result = await _taskService.AcceptTask(connectionId, taskInstanceId);

        if (result.IsSuccess)
            return new MessageResult("task_accepted", connectionId, new { task = result.Value });

        return new MessageResult("error", connectionId, new { message = result.Error ?? "Could not accept task" });
    }

    private async Task<IActionResult> CompleteTask(string connectionId, JsonElement payload)
    {
        string taskInstanceId;
        try
        {
            taskInstanceId = payload.ReqStr("task_instance_id");
        }
        catch
        {
            return new MessageResult("error", connectionId, new { message = "Task instance ID required" });
        }

        var result = await _taskService.CompleteTask(connectionId, taskInstanceId);

        if (result.IsSuccess)
        {
            var snapshot = _playerService.GetSnapshot(connectionId);
            return new MessageResult("task_completed", connectionId, new
            {
                task = result.Value is null ? null : new Dictionary<string, object?>
                {
                    ["taskId"] = result.Value.TaskId,
                    ["instanceId"] = result.Value.InstanceId,
                    ["title"] = result.Value.Title,
                    ["status"] = result.Value.Status.ToString()
                },
                player = snapshot?.SnapshotToMap()
            });
        }

        return new MessageResult("error", connectionId, new { message = result.Error ?? "Could not complete task" });
    }
}
